drivers/gpio/gpio-siklu-cpld.c

```diff
@@ -0,0 +1,188 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include <linux/module.h>
+#include <linux/mutex.h>
+#include <linux/of_platform.h>
+#include <linux/platform_device.h>
+#include <linux/gpio/driver.h>
+#include <linux/firmware/siklu-cpld.h>
+
+#define DRV_NAME "siklu-cpld-gpio"
+
+static DEFINE_MUTEX(siklu_cpld_gpio_mutex);
+
+/* offset values: value/bits 0-7 for reg 0x7, value/bits 8-15 for reg 0xf, value/bits 16-23 for reg 0x14 */
+static int siklu_cpld_gpio_get(struct gpio_chip *gc, unsigned offset)
+{
+	struct device *cpld_dev = gpiochip_get_data(gc);
+	u8 reg_val, cpld_reg = R_CPLD_LOGIC_SFP_MODE;
+	unsigned adjust_offset = 0;
+
+	if (offset <= 7) {
+		cpld_reg = R_CPLD_LOGIC_SFP_MODE;
+		adjust_offset = offset;
+	} else if ((offset >= 8) && (offset <= 15)) {
+		cpld_reg = R_CPLD_LOGIC_INT_HNDLR_1;
+		adjust_offset = offset - 8;
+	} else {
+		cpld_reg = R_CPLD_LOGIC_MISC_1;
+		adjust_offset = offset - 16;
+	}
+
+	siklu_cpld_reg_read(cpld_dev, cpld_reg, &reg_val);
+
+/*	printk(KERN_ERR
+	       "siklu_cpld_gpio_get, offset = 0x%x, cpld_reg = 0x%x, reg_val = 0x%x \n",
+	       offset, cpld_reg, reg_val);*/
+
+	return !!(reg_val & (1 << adjust_offset));
+#if 0	
+	struct device *cpld_dev = gpiochip_get_data(gc);
+	u8 reg_val;
+
+	siklu_cpld_reg_read(cpld_dev, R_CPLD_LOGIC_GPIO, &reg_val);
+
+	printk(KERN_ERR
+	       "siklu_cpld_gpio_get, offset = 0x%x, reg_val = 0x%x \n",
+	       offset, reg_val);
+
+	return !!(reg_val & (1 << offset));
+#endif
+}
+
+/* offset values: value/bits 0-7 for reg 0x7, value/bits 8-15 for reg 0xf, value/bits 16-23 for reg 0x14 */
+static void siklu_cpld_gpio_set(struct gpio_chip *gc, unsigned offset, int value)
+{
+	struct device *cpld_dev = gpiochip_get_data(gc);
+	u8 reg_val, cpld_reg = R_CPLD_LOGIC_SFP_MODE;
+	unsigned adjust_offset = 0;
+
+	if (offset <= 7)
+	{
+		cpld_reg = R_CPLD_LOGIC_SFP_MODE;
+		adjust_offset = offset;
+	}
+	else if ((offset >= 8) && (offset <= 15))
+	{
+		cpld_reg = R_CPLD_LOGIC_INT_HNDLR_1;
+		adjust_offset = offset - 8;
+	}
+	else
+	{
+		cpld_reg = R_CPLD_LOGIC_MISC_1;
+		adjust_offset = offset - 16;
+	}
+
+	mutex_lock(&siklu_cpld_gpio_mutex);
```

Comment:
no need to lock it in siklu_cpld_gpio_get?

Reply:
I don't know what to tell you. This code is based on what Baruch Tikos wrote. And he wrote in both "siklu_cpld_gpio_set" and "siklu_cpld_gpio_get". I don't know what problems he saw and decided to do it also here.